File: ClassArray.java
import java.io.Serializable;
import java.util.Dictionary;
import java.util.Hashtable;

public class ClassArray<T extends ITechno> implements Serializable {

	private Dictionary<Integer, T> places;

	int maxCount;

	private T defaultValue;

	public ClassArray(int size, T defVal)

	{

		defaultValue = defVal;

		places = new Hashtable<Integer, T>();

		maxCount = size;

	}

	public static <T extends ITechno> int plus(ClassArray<T> p, T plane)
			throws PortOverflowException

	{

		if (p.places.size() == p.maxCount)
			throw new PortOverflowException();

		for (int i = 0; i < p.places.size(); i++)

		{

			if (p.checkFree(i))

			{

				p.places.put(i, plane);

				return i;

			}

		}

		p.places.put(p.places.size(), plane);

		return p.places.size() - 1;

	}

	public static <T extends ITechno> T minus(ClassArray<T> p, int index)
			throws PortIndexOutOfRangeException

	{

		if (p.places.get(index) != null)

		{

			T plane = p.places.get(index);

			p.places.remove(index);

			return plane;

		}

		throw new PortIndexOutOfRangeException();

	}

	public boolean checkFree(int index)

	{

		if (places.get(index) == null)
			return true;

		return false;

	}

	public T getPlane(int ind) {

		if (places.get(ind) != null)
			return places.get(ind);

		return defaultValue;

	}

}

File: ClassArray.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApplication3
{
    class ClassArray<T> where T : ITechno
    {
        private T defaultValue;
        private Dictionary<int, T> places;
        private int maxCount;
        
        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            places = new Dictionary<int, T>();
            maxCount = size;            
        }
      
        public static int operator +(ClassArray<T> p, T ship)
        {
            if (p.places.Count == p.maxCount)
            {
                return -1;
            }
            for (int i = 0; i < p.places.Count; i++)
            {
                if (p.CheckFreePlace(i))
                {
                    p.places.Add(i, ship);
                    return i;
                }
            }

            p.places.Add(p.places.Count, ship);
            return p.places.Count - 1;
          
        }
        public static T operator -(ClassArray<T> p, int index)
        {
            if (p.places.ContainsKey(index))
            {
                T ship = p.places[index];
                p.places.Remove(index);
                return ship;
            }
            return p.defaultValue;
        }
        private bool CheckFreePlace(int index)
        {
            return !places.ContainsKey(index);
        }

        public T this[int ind]
        {
            get
            {
                if (places.ContainsKey(ind))
                {
                    return places[ind];

                }
                return defaultValue;
            }
        }


    }
}

